### PR TITLE
fix: Capture more details of ppom errors

### PR DESCRIPTION
**Pull request by segun**
## **Description**

We're seeing ppom throwing errors on Flask, but we don't have further details on what the error is. We should capture more details about the error. 

This solution leverages both Sentry or Mixpanel. we are passing the error as the value for `security_alert_reason` property at the same time sending the error to sentry.

## **Related issues**

Fixes: [#1535](https://github.com/MetaMask/MetaMask-planning/issues/1535)

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: app/scripts/lib/transaction/util.ts [hunk 1]
```diff
@@ -111,7 +112,16 @@ export async function addTransaction(
...
+          try {
+            return ppom.validateJsonRpc(ppomRequest);
+          } catch (error) {
+            // send to sentry
+            captureException(error);
+            return {
+              reason: error,
+              result_type: BlockaidResultType.Errored,
+            };
+          }
```
Comment (jpuri) [inline app/scripts/lib/transaction/util.ts:124]: Can you plz check with product team is its ok to have a value for reason apart from those predefined for it.

Comment (bschorchit) [reply app/scripts/lib/transaction/util.ts:124]: I've confirmed it's with Olu :) thank you for suggesting this check!

Comment (segun) [reply app/scripts/lib/transaction/util.ts:124]: I spoke with @bschorchit she said it's fine to have the error in the reason.

This is the description/proposed solution in the issue

```
We're seeing ppom throwing errors on Flask, but we don't have further details on what the error is. We should capture more details about the error.

As discussed previously we can leverage either Sentry or Mixpanel for that. If we leverage Sentry, we should make sure that these errors do appear on Sentry and are not capped out due to Sentry hitting it's error limit. If we choose Mixpanel, we can pass the error as the value for security_alert_reason property or add a new property to the events to pass the error. We could also do both Sentry and Mixpanel.
```

**Review by jpuri — DISMISSED** [review]

File: app/scripts/lib/ppom/ppom-middleware.ts [hunk 2]
```diff
@@ -64,7 +65,17 @@ export function createPPOMMiddleware(
         // eslint-disable-next-line require-atomic-updates
         req.securityAlertResponse = await ppomController.usePPOM(
           async (ppom: PPOM) => {
-            return ppom.validateJsonRpc(req);
+            try {
+              return ppom.validateJsonRpc(req);
+            } catch (error) {
+              // send to sentry
```
Comment (digiwand) [inline app/scripts/lib/ppom/ppom-middleware.ts:71]:
```suggestion
```

I think importing from the sentry library may imply this

File: shared/constants/security-provider.ts [hunk 3]
```diff
@@ -55,6 +55,7 @@ export enum BlockaidResultType {
   Malicious = 'Malicious',
   Warning = 'Warning',
   Benign = 'Benign',
+  Errored = 'Errored',
```
Comment (digiwand) [inline shared/constants/security-provider.ts:58]: based on the [internal thread](https://consensys.slack.com/archives/C05HZHDS89Z/p1706541505600509?thread_ts=1705393852.556679&cid=C05HZHDS89Z), I believe we will want to use "Error" as the value

**Review by digiwand — CHANGES_REQUESTED** [review]
although this value is for our internal Sentry, I believe it would be best to stay consistent with Blockaid's new result_type value name "Error" rather than using "Errored"

File: shared/constants/security-provider.ts (hunk 3 above)
Comment (digiwand) [reply shared/constants/security-provider.ts:58]: using `Errored = 'Error'` in https://github.com/MetaMask/metamask-extension/pull/22742

**Comment by segun** [comment]
@danjm merged a PR that fixes the issues in this PR, @bschorchit @jpuri confirm it's okay to close this issue?

File: shared/constants/security-provider.ts [hunk 4]
```diff
@@ -55,10 +55,11 @@ export enum BlockaidResultType {
   Malicious = 'Malicious',
   Warning = 'Warning',
   Benign = 'Benign',
-
   // MetaMask defined result types
   Failed = 'Failed',
   NotApplicable = 'NotApplicable',
+  /* eslint-disable-next-line @typescript-eslint/no-shadow */
+  Error = 'Error',
```
Comment (digiwand) [inline shared/constants/security-provider.ts:62]: when we resolve the merge conflict it should bring in this change
```suggestion
  Errored = 'Error',
```
removing the eslint disable line

will need to update the other instances in this PR with it

File: test/e2e/helpers.js [hunk 5]
```diff
@@ -719,6 +719,24 @@ const sendScreenToConfirmScreen = async (
...
+          driver.clickElement({
+            text: 'I understand',
+            tag: 'button',
+          });
+        },
+        (error) => {
+          console.error('Element not found.', error);
+        },
+      );
+
```
Comment (jpuri) [inline test/e2e/helpers.js:739]: @segun : is above related change ?

File: app/scripts/lib/createRPCMethodTrackingMiddleware.js [hunk 6]
```diff
@@ -212,6 +212,14 @@ export default function createRPCMethodTrackingMiddleware({
           BlockaidResultType.NotApplicable;
         eventProperties.security_alert_reason =
           req.securityAlertResponse?.reason ?? BlockaidReason.notApplicable;
+
+        if (
+          req.securityAlertResponse?.reason === BlockaidReason.failed &&
```
Comment (segun) [inline app/scripts/lib/createRPCMethodTrackingMiddleware.js:217]:
```suggestion
          req.securityAlertResponse?.result_type === BlockaidResulttype.Errored &&
```

Comment (segun) [reply app/scripts/lib/createRPCMethodTrackingMiddleware.js:217]: refactor BlockaidReason.failed to BlockaidReason.error

File: shared/constants/metametrics.ts [hunk 7]
```diff
@@ -766,6 +766,7 @@ export enum MetaMetricsEventUiCustomization {
   FlaggedAsWarning = 'flagged_as_warning',
   GasEstimationFailed = 'gas_estimation_failed',
   SecurityAlertFailed = 'security_alert_failed',
```
Comment (segun) [inline shared/constants/metametrics.ts:768]:
```suggestion

```

Comment (segun) [reply shared/constants/metametrics.ts:768]: Check that it's not used anywhere else

**Review by segun — COMMENTED** [review]
FYI

**Review by jpuri — APPROVED** [review]

**Review by blackdevelopa — APPROVED** [review]

**Review by digiwand — APPROVED** [review]

**Comment by digiwand** [comment]
no blockers. If the comments above sound good to you too, we can add it in a follow-up PR

**Pull request merged**